### PR TITLE
[main][feat] Add example on new position for BUSD CakeMaxi

### PR DESCRIPTION
**Pull request by ricardo-alpaca**
## Description
Add example on new position for BUSD CakeMaxi

### CODE REVIEW

File: example/open-new-position.ts [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+  // this could proven to be dangerous for price manipulation attack, we put it as `0` here for the simplicity of the code example
+  // Alpaca Finance frontend would calculate this value based on the current price with some percentage of slippage tolerance
+  const minFarmingToken = '0'
+
+  // In this example, we are opening a new position on Cake Single Staking Farm with BUSD as borrowing asset
+  // We will farm this position with our 100 BUSD principal and we will borrow 100 BUSD from Alpaca Finance's Lending Vault
+  // We do not supply any farming token (CAKE in this case) here.
+  const inputBaseTokenAmount = ethers.utils.parseEther('100')
+  const inputFarmingTokenAmount = ethers.utils.parseEther('0')
+  const loanAmount = ethers.utils.parseEther('100')
```
Comment (spicysquid168) [inline example/open-new-position.ts:42]: borrowAmount

File: hardhat.config.ts [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+        {
+          privateKey: process.env.LOCAL_PRIVATE_KEY_4,
+          balance: "10000000000000000000000",
+        },
+      ],
+    },
+    testnet: {
+      url: "https://data-seed-prebsc-1-s3.binance.org:8545",
+      accounts: [process.env.BSC_TESTNET_PRIVATE_KEY],
+    },
```
Comment (almost-llamas) [inline hardhat.config.ts:38]: Let's remove testnet and hardhat network, and set mainnet as default

File: hardhat.config.ts [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+  paths: {
+    sources: "./contracts/6",
+    tests: "./test",
+    cache: "./cache",
+    artifacts: "./artifacts",
+  },
+  typechain: {
+    outDir: "./typechain",
+    target: process.env.TYPECHAIN_TARGET || "ethers-v5",
+  },
```
Comment (almost-llamas) [inline hardhat.config.ts:82]: We could lean this, since we don't need compilation.

File: example/open-new-position.ts [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+  )
+  if (baseTokenAllowance.isZero()) {
+    await baseToken.increaseAllowance(vaultAddress, constants.MaxUint256)
+  }
+  const farmingTokenAllowance = await farmingToken.allowance(
+    signer.address,
+    vaultAddress,
+  )
+  if (!inputFarmingTokenAmount.isZero() && farmingTokenAllowance.isZero()) {
+    await farmingToken.increaseAllowance(vaultAddress, constants.MaxUint256)
```
Comment (spicysquid168) [inline example/open-new-position.ts:72]: `.approve` is better. Not all BEP20 support `increaseAllowance`

File: example/open-new-position-lp.ts [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+ // _____                               _                
+ // |  __ \                             | |               
+ // | |__) |_ _ _ __ __ _ _ __ ___   ___| |_ ___ _ __ ___ 
+ // |  ___/ _` | '__/ _` | '_ ` _ \ / _ \ __/ _ \ '__/ __|
+ // | |  | (_| | | | (_| | | | | | |  __/ ||  __/ |  \__ \
+ // |_|   \__,_|_|  \__,_|_| |_| |_|\___|\__\___|_|  |___/
+                                                        
+  const vaultAddress = '0xe5ed8148fE4915cE857FC648b9BdEF8Bb9491Fa5' // BUSD Vault
+  const workerAddress = '0x8164104FeaA27Ac52d7CD22D63a5Ef50971cB7ff' // ALPACA-BUSD PancakeswapWorker
+  const strategyAddress = '0x22fC6110d5d9b122f3C2d4715C24566342161a12' // ALPACA-BUSD PancakeswapWorker.strategies.StrategyAddTwoSidesOptimal
```
Comment (zhfk930129) [inline example/open-new-position-lp.ts:29]: Is this address for [PancakeswapV2StrategyAddTwoSidesOptimal](https://github.com/alpaca-finance/bsc-alpaca-contract/blob/main/contracts/6/protocol/strategies/pancakeswapV2/PancakeswapV2StrategyAddTwoSidesOptimal.sol) or for [PancakeswapV2RestrictedStrategyAddTwosidesOptimal](https://github.com/alpaca-finance/bsc-alpaca-contract/blob/main/contracts/6/protocol/strategies/pancakeswapV2-restricted/PancakeswapV2RestrictedStrategyAddTwosidesOptimal.sol)?

Comment (zhfk930129) [reply example/open-new-position-lp.ts:29]: I just opened a real position using [Alpaca Finance frontend on testnet](https://testnet-lfg-app.alpacafinance.org/farm) and here is the [transaction](https://testnet.bscscan.com/tx/0xc7beeab4ee89fadca4d0684bb3fbcf1c6dc44cc2a0507a21688fc787a70b3b7c).

From the transaction input data it seems the strategy is still the old one: PancakeswapV2StrategyAddTwoSidesOptimal. See input [8] and [9], those should be two addresses, right?
image

Is your test case working now?

Comment (zhfk930129) [reply example/open-new-position-lp.ts:29]: If this transaction is executing on the new strategy, baseTokenAddress and farmingTokenAddress should not be here... Then what does [8] and [9] mean?

File: example/open-new-position-lp.ts [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+    '0',
+    ethers.utils.defaultAbiCoder.encode(
+      ['address', 'bytes'],
+      [strategyAddress, encodedStrategyParams],
+    ),
+    {
+      gasLimit: OpenPositionGasLimit,
+      value: '0',
+    },
+  )
```
Comment (zhfk930129) [inline example/open-new-position-lp.ts:100]: Sorry you are using js / ts library to interact with Alpaca contract here. While our requirement is to use contract to contract interaction.

Will the following code work properly to open a new position?
```javascript
alpacaVault.work(
    0,
    alpacaWorkerAddress,
    principalAmount,
    borrowAmount,
    0,
    abi.encode(
        alpacaStrategyAddress,
        abi.encode(
            farmingAmount,
            minimumLPAmount
        )
    )
);
```

File: example/open-new-position-lp.ts (hunk 5 above)
Comment (ricardo-alpaca) [reply example/open-new-position-lp.ts:29]: > Is this address for [PancakeswapV2StrategyAddTwoSidesOptimal](https://github.com/alpaca-finance/bsc-alpaca-contract/blob/main/contracts/6/protocol/strategies/pancakeswapV2/PancakeswapV2StrategyAddTwoSidesOptimal.sol) or for [PancakeswapV2RestrictedStrategyAddTwosidesOptimal](https://github.com/alpaca-finance/bsc-alpaca-contract/blob/main/contracts/6/protocol/strategies/pancakeswapV2-restricted/PancakeswapV2RestrictedStrategyAddTwosidesOptimal.sol)?

This is the address for PancakeswapV2RestrictedStrategyAddTwosidesOptimal. I have verified the contract in testnet which you can view the implementation code here https://testnet.bscscan.com/address/0xda6031c074d41a28df8827b3cf6b66352ea3ba7a#readProxyContract

Comment (ricardo-alpaca) [reply example/open-new-position-lp.ts:29]: > If this transaction is executing on the new strategy, baseTokenAddress and farmingTokenAddress should not be here... Then what does [8] and [9] mean?

[8] and [9] seem to be just a placeholder of some sort. I am not sure about this but they do not matter in our context. The ones you need to look for are [10] and [11] instead.

[10] is `farmingTokenAmount` where the value `3635c9adc5dea00000` in base16 equals to `1000000000000000000000` or 100 BUSD which is consistent with the 100 BUSD transfer in the transaction.

[11] is `minLpAmount` which is a slippage control calculated from the frontend. The value `1c4e918b51a1a28389` in base16 equals to `522170293453169263497` or 522.1702934531693 minimum LP token. This transaction passed the `minLpAmount` check by producing 523.479651593504030007 LP token which is larger than the `minLpAmount`.

Comment (zhfk930129) [reply example/open-new-position-lp.ts:29]: Yes thanks to explain! Our understanding on [10] and [11] are the same.

File: example/open-new-position-lp.ts (hunk 6 above)
Comment (ricardo-alpaca) [reply example/open-new-position-lp.ts:100]: Something along that line should work. I'll try doing a test with contract to contract call.

**Pull request merged**